sklearn/datasets/tests/test_base.py

```diff
@@ -363,3 +367,24 @@ def test_load_boston_error():
     msg = "cannot import name 'non_existing_function' from 'sklearn.datasets'"
     with pytest.raises(ImportError, match=msg):
         from sklearn.datasets import non_existing_function  # noqa
+
+
+def test_fetch_remote_raise_warnings_with_invalid_url(monkeypatch):
+    """Check retry mechanism in _fetch_remote."""
+
+    url = "https://scikit-learn.org/this_file_does_not_exist.tar.gz"
+    invalid_remote_file = RemoteFileMetadata("invalid_file", url, None)
+    urlretrieve_mock = Mock(
+        side_effect=HTTPError(url=url, code=404, msg="Not Found", hdrs=None, fp=None)
```

Comment:
There is an underlying issue with `HttpError` and some Python versions. We can work around this by:

```suggestion
        side_effect=HTTPError(url=url, code=404, msg="Not Found", hdrs=None, fp=io.BytesIO())
```

and `import io` at the top of the file.